### PR TITLE
util: Report parse errors in configuration file

### PR DESCRIPTION
**Pull request by laanwj**
Report errors while parsing the configuration file, instead of silently ignoring them.

    $ src/bitcoind -regtest
    Error reading configuration file: parse error on line 22: nodebuglogfile, if you intended to specify a negated option, use nodebuglogfile=1 instead
    $ src/bitcoind -regtest
    Error reading configuration file: parse error on line 22: sdafsdfafs
    $ src/bitcoind -regtest
    Error reading configuration file: parse error on line 24: -nodebuglogfile=1, options in the configuration file must be specified without leading -

(inspired by https://github.com/bitcoin/bitcoin/pull/14100#issuecomment-417264823)

### CODE REVIEW

File: src/util.cpp [hunk 1]
```diff
@@ -830,17 +830,27 @@ static std::vector<std::pair<std::string, std::string>> GetConfigOptions(std::is
                 std::string name = prefix + TrimString(str.substr(0, pos), pattern);
                 std::string value = TrimString(str.substr(pos + 1), pattern);
                 options.emplace_back(name, value);
+            } else {
+                error = strprintf("parse error on line %i: %s", linenr, str);
+                if (str.size() >= 2 && str.substr(0, 2) == "no") {
+                    error += strprintf(", if you intended to specify a negated option, use %s=1 instead", str);
```
Comment (laanwj) [inline src/util.cpp]: we could also change this logic here to interpret it as `noX=1`, but I'm not sure what kind of syntax we want to support

**Comment by practicalswift** [comment]
Concept ACK

Nice usability improvement!

File: src/util.cpp (hunk 1 above)
Comment (Nathapon8899) [reply src/util.cpp]: error = strprintf ( "แยกวิเคราะห์ข้อผิดพลาดในบรรทัด% i:% s " , linenr, str);
                if (str. size ()> = 2 && str. substr ( 0 , 2 ) == " noX=1 " ) {
                    ข้อผิดพลาด + = strprintf ( "ถ้าคุณตั้งใจจะระบุตัวเลือกที่ถูกละเลยให้ใช้% s = 1 แทน" , str);

**Comment by laanwj** [comment]
updated to also check for stray `-` in config file

**Comment by ken2812221** [comment]
utACK a66c0f78a941968340f030911765a84219908c4d

**Comment by MarcoFalke** [comment]
Could add some tests, otherwise looks good:


```diff
diff --git a/test/functional/feature_config_args.py b/test/functional/feature_config_args.py
index 62091048f9..9be59b32b4 100755
--- a/test/functional/feature_config_args.py
+++ b/test/functional/feature_config_args.py
@@ -14,8 +14,29 @@ class ConfArgsTest(BitcoinTestFramework):
         self.setup_clean_chain = True
         self.num_nodes = 1
 
+    def test_config_file_parser(self):
+        # Assume node is stopped
+
+        inc_conf_file_path = os.path.join(self.nodes[0].datadir, 'include.conf')
+        with open(os.path.join(self.nodes[0].datadir, 'bitcoin.conf'), 'a', encoding='utf-8') as conf:
+            conf.write('includeconf={}\n'.format(inc_conf_file_path))
+
+        with open(inc_conf_file_path, 'w', encoding='utf-8') as conf:
+            conf.write('-dash=1\n')
+        self.nodes[0].assert_start_raises_init_error(expected_msg='Error reading configuration file: parse error on line 1: -dash=1, options in configuration file must be specified without leading -')
+
+        with open(inc_conf_file_path, 'w', encoding='utf-8') as conf:
+            conf.write('nono\n')
+        self.nodes[0].assert_start_raises_init_error(expected_msg='Error reading configuration file: parse error on line 1: nono, if you intended to specify a negated option, use nono=1 instead')
+
+        with open(inc_conf_file_path, 'w', encoding='utf-8') as conf:
+            conf.write('')  # clear
+
     def run_test(self):
         self.stop_node(0)
+
+        self.test_config_file_parser()
+
         # Remove the -datadir argument so it doesn't override the config file
         self.nodes[0].args = [arg for arg in self.nodes[0].args if not arg.startswith("-datadir")]

**Comment by laanwj** [comment]
@MarcoFalke thanks, that's a neat test, added commit

**Comment by MarcoFalke** [comment]
Backported in eb202ea21de1fbcc32fb3db63fcce4135b7a2849 and 83aafd5

**Pull request merged**